dip-template/runtimes/dip-receiver/src/dip.rs

```diff
@@ -31,3 +37,90 @@ impl pallet_dip_receiver::Config for Runtime {
 	type RuntimeEvent = RuntimeEvent;
 	type RuntimeOrigin = RuntimeOrigin;
 }
+
+fn derive_verification_key_relationship(call: &RuntimeCall) -> Option<DidVerificationKeyRelationship> {
+	match call {
+		RuntimeCall::DidLookup { .. } => Some(DidVerificationKeyRelationship::Authentication),
+		RuntimeCall::Utility(pallet_utility::Call::batch { calls }) => single_key_relationship(calls).ok(),
+		RuntimeCall::Utility(pallet_utility::Call::batch_all { calls }) => single_key_relationship(calls).ok(),
+		RuntimeCall::Utility(pallet_utility::Call::force_batch { calls }) => single_key_relationship(calls).ok(),
+		_ => None,
+	}
+}
+
+// Taken and adapted from `impl
+// did::DeriveDidCallAuthorizationVerificationKeyRelationship for RuntimeCall`
+// in Spiritnet/Peregrine runtime.
+fn single_key_relationship(calls: &[RuntimeCall]) -> Result<DidVerificationKeyRelationship, ()> {
+	let first_call_relationship = calls.get(0).and_then(derive_verification_key_relationship).ok_or(())?;
+	calls
+		.iter()
+		.skip(1)
+		.map(derive_verification_key_relationship)
+		.try_fold(first_call_relationship, |acc, next| {
+			if next == Some(acc) {
+				Ok(acc)
+			} else {
+				Err(())
+			}
+		})
```

Comment:
FYI This could get simplified with reduce once it's not experimental... https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_reduce